Automattic/wp-calypso
# composite-checkout: Move stripe wrapper to host page

#### Changes proposed in this Pull Request

This PR changes how payment methods load Stripe.js by forcing the host page to load it rather than handling the loading themselves. The motivation was so that we can check for Apple Pay availability before loading any payment methods, but it has the added advantage of simplifying the API of payment methods. 

Previously each payment method object could have a `checkoutWrapper` property which would be used to wrap the entirety of the checkout form. Although this could have other applications the intent was to make it possible for payment methods that required a shared stripe context to create one in state that was shared between sibling components in the form.

With this PR, we remove the concept of a `checkoutWrapper` and instead simply make it a requirement of stripe payment methods (currently credit card and apple pay) that the host page load, configure, initialize, and pass the `stripe` instance into the payment methods itself. This is actually quite a lot more flexible since the host page may want to load stripe in any number of ways and it prevents the possibility of loading stripe more than once on the same checkout.

With that done, this PR creates a hook `useIsApplePayAvailable` which it uses to determine if it should add the apple pay method.

One side effect of this change is that we no longer need to duplicate the stripe library code in the `composite-checkout` package; we can just use the version already on the host page (calypso). (The duplicate library still remains in the package but it is only used for the demo.)

#### Testing instructions

- Sandbox the store.
- First ignore apple pay and try visiting composite checkout in a non-Safari browser. Make sure that the Stripe payment method loads correctly.
- Using Mac OS Safari, visit WordPress.com and log in. This is required because once we proxy, you will not be able to log in.
- Enable Charles proxy to spoof WordPress.com. This is required because Apple Pay will only work on SSL and on a production domain. The details of how to do this are beyond the scope of this PR. Contact me if you have any issues.
- Using Mac OS Safari, reload WordPress.com and be sure you see the "dev" badge in the lower-right.
- Add a plan to your cart and visit checkout.
- Select "Apple Pay" from the available payment methods.
- Complete the form and press "Pay". You must use a real Apple Pay card. But as long as the store is sandboxed, you won't actually be charged.
- Complete the Apple Pay sheet that appears.
- Verify that the payment was successful and that the plan was added to your site.

#### Screenshots

![Screen Shot 2020-02-25 at 12 20 46 AM](https://user-images.githubusercontent.com/2036909/75217335-c152da00-5764-11ea-987e-772d4dbc7dee.png)

![Screen Shot 2020-02-25 at 12 20 55 AM](https://user-images.githubusercontent.com/2036909/75217337-c44dca80-5764-11ea-9cba-f7ec2e3c192b.png)
